lib/shore/client/tokens/v1/merchant_account.rb

```diff
@@ -0,0 +1,151 @@
+module Shore
+  module Client
+    module Tokens
+      module V1
+        class MerchantAccount # :nodoc:
+          TYPE = 'merchant-account'.freeze
+          attr_reader :id, :type, :version
+
+          # Builds a hash where the Merchant UUID is the key and the Role is the
+          # value.
+          # @param merchant_uuids [Array<UUID>]
+          # @return Hash<UUID,String>
+          def roles(merchant_uuids)
+            {}.tap do |h|
+              merchant_uuids.each do |merchant_uuid|
+                h[merchant_uuid] = role(merchant_uuid)
+              end
+            end
+          end
+
+          # Returns a role the merchant belongs to as a string or nil of the
+          # MerchantAccount does not have access to the merchant.
+          # @param merchant_uuid [UUID] Merchant UUID
+          # @return [String]
+          def role(merchant_uuid)
+            return nil unless include?(merchant_uuid)
+
+            return 'member' if member?(merchant_uuid)
+            return 'admin' if admin?(merchant_uuid)
+            return 'owner' if owner?(merchant_uuid)
+            return nil
+          end
+
+          # Just for backwards compatibility in CORE.
+          def import_roles(role_entries)
+            role_entries.each do |role_entry|
+              role_entry = role_entry.with_indifferent_access
+
+              name = role_entry[:role]
+
+              next if name.nil?
+
+              @roles[name.to_sym] << role_entry[:id]
+            end
+          end
+
+          # @param id [UUID] - MerchantAccount uuid.
+          # @param attributes [Hash] - JWT data.
+          def initialize(id, attributes = {})
+            @id = id
+            @type = TYPE
+            @version = 1
+            @roles = {}
+
+            @roles[:owner] = build_uuids(attributes[:owner] || [])
+            @roles[:admin] = build_uuids(attributes[:admin] || [])
+            @roles[:member] = build_uuids(attributes[:member] || [])
+
+            import_roles(attributes[:roles]) if attributes[:roles]
+          end
+
+          def build_uuids(urlsafe_uuids = [])
+            urlsafe_uuids.map { |uuid| UUID.from_urlsafe(uuid) }
```

Comment:
If you use a ruby Set, then lookups in that set use constant time.
```
Set.new(urlsafe_uuids.map { |uuid| UUID.from_urlsafe(uuid) })
```
The following code then has O(1) (i.e. constant time):
```
owners.include?(merchant_uuid)
```